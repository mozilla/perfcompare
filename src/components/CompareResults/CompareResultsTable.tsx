import Alert from '@mui/material/Alert';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import Grid from '@mui/material/Grid';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableContainer from '@mui/material/TableContainer';

import { RootState } from '../../common/store';
import { useAppSelector } from '../../hooks/app';
import type { CompareResultsState } from '../../types/state';
import CompareTableStatus from './CompareTableStatus';
import PaginatedCompareResults from './PaginatedCompareResults';

function CompareResultsTable(props: CompareResultsProps) {
  const { theme } = props;
  const compareResults: CompareResultsState = useAppSelector(
    (state: RootState) => state.compareResults,
  );

  return (
    <>
      {!compareResults.loading && (
        <>
          <CompareTableStatus />
          <TableContainer component={Paper}>
            <Table
              sx={{ minWidth: 650 }}
              size="small"
              aria-label="a dense table"
            >
              <colgroup>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'20%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'10%' }}/>
      <col style={{ width:'12%' }}/>
      <col style={{ width:'8%' }}/>
   </colgroup>
              <PaginatedCompareResults theme={theme} />
            </Table>
          </TableContainer>
        </>
      )}
      {compareResults.loading && (
        <Box display="flex" justifyContent="center">
          <Grid item xs={1} className="compare-results-error">
            <CircularProgress />
          </Grid>
        </Box>
      )}
      {compareResults.error && (
        <Box display="flex" justifyContent="center">
          <Grid item xs={6} className="compare-results-error">
            <Alert severity="error" sx={{ textAlign: 'center' }}>
              Error: {compareResults.error}
            </Alert>
          </Grid>
        </Box>
      )}
    </>
  );
}

interface CompareResultsProps {
  theme: string;
}

export default CompareResultsTable;
