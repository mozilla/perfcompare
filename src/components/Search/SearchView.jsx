import React from 'react';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';

import SearchInput from './SearchInput';

export default function SearchView() {
  return (
    <Container maxWidth="lg">
      <Box>
        <Typography variant="h1" component="div" align="center" gutterBottom>
          PerfCompare
        </Typography>
      </Box>
      <SearchInput />
    </Container>
  );
}
