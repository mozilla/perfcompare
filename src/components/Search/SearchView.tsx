import ArrowForward from '@mui/icons-material/ArrowForward';
import Button from '@mui/material/Button';
import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import { useSnackbar, VariantType } from 'notistack';
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import { repoMap, featureNotSupportedError } from '../../common/constants';
import type { RootState } from '../../common/store';
import useFilterCompareResults from '../../hooks/useFilterCompareResults';
import { Strings } from '../../resources/Strings';
import { Revision } from '../../types/state';
import RevisionSearch from '../Shared/RevisionSearch';
import SelectedRevisionsTable from '../Shared/SelectedRevisionsTable';
import SearchViewInit from './SearchViewInit';

const strings = Strings.components.compareDefault.sharedCollasped;

function SearchView(props: SearchViewProps) {
  const navigate = useNavigate();
  const { clearFilters } = useFilterCompareResults();
  const { enqueueSnackbar } = useSnackbar();
  const warningVariant: VariantType = 'warning';

  const goToCompareResultsPage = (selectedRevisions: Revision[]) => {
    // TODO: remove this check once comparing without a base
    //  and comparing multiple revisions against a base is enabled
    if (selectedRevisions.length === 1 || selectedRevisions.length > 2) {
      enqueueSnackbar(featureNotSupportedError as string, {
        variant: warningVariant,
      });
      return;
    }
    const revs = selectedRevisions.map((rev) => rev.revision);
    const repos = selectedRevisions.map((rev) => repoMap[rev.repository_id]);
    clearFilters();
    navigate({
      pathname: '/compare-results',
      search: `?revs=${revs.join(',')}&repos=${repos.join(',')}`,
    });
  };

  const { selectedRevisions } = props;

  return (
    <Container className='perfcompare-body' maxWidth='lg'>
      {/* Component to fetch recent revisions on mount */}
      <SearchViewInit />
      <Grid item xs={12}>
        {selectedRevisions.length > 0 && (
          <SelectedRevisionsTable view='search' />
        )}
      </Grid>
      <Grid item className='compare-button-section'>
        {selectedRevisions.length > 0 && (
          <Button
            className='compare-button'
            variant='contained'
            onClick={() => goToCompareResultsPage(selectedRevisions)}
          >
            {strings.button}
            <ArrowForward className='compare-icon' />
          </Button>
        )}
      </Grid>
      <RevisionSearch view='search' />
    </Container>
  );
}

interface SearchViewProps {
  searchResults: Revision[];
  selectedRevisions: Revision[];
}

function mapStateToProps(state: RootState) {
  return {
    searchResults: state.search.searchResults,
    selectedRevisions: state.selectedRevisions.revisions,
  };
}

export default connect(mapStateToProps)(SearchView);
