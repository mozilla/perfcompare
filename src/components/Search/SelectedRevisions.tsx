import { useEffect, useState } from 'react';

import Box from '@mui/material/Box';
import List from '@mui/material/List';

import { repoMap, searchView } from '../../common/constants';
import { useAppSelector } from '../../hooks/app';
import { SelectRevsStyles } from '../../styles';
import {
  InputType,
  View,
  RevisionsList,
  Repository,
  ThemeMode,
} from '../../types/state';
import SelectedRevisionItem from './SelectedRevisionItem';

interface SelectedRevisionsProps {
  mode: ThemeMode;
  searchType: InputType;
  isWarning: boolean;
  view: View;
  editBtnVisible?: boolean;
}

function SelectedRevisions({
  mode,
  searchType,
  isWarning,
  view,
  editBtnVisible,
}: SelectedRevisionsProps) {
  const styles = SelectRevsStyles(mode);
  const [revisions, setRevisions] = useState<RevisionsList[]>([]);
  const [repositories, setRepositories] = useState<Repository['name'][]>([]);
  const checkedRevisionsList = useAppSelector(
    (state) => state.search[searchType].checkedRevisions,
  );

  const selectedRevisions = useAppSelector(
    (state) => state.selectedRevisions.revisions,
  );

  const displayedSelectedRevisions = useAppSelector(
    (state) => state.selectedRevisions[searchType],
  );

  const checkedRepositories = checkedRevisionsList.map((item) => {
    const selectedRep = repoMap[item.repository_id];
    return selectedRep;
  });

  const selectedRevRepo = selectedRevisions.map((item) => {
    const selectedRep = repoMap[item.repository_id];
    return selectedRep;
  });

  useEffect(() => {
    if (view === searchView || !editBtnVisible) {
      setRevisions(checkedRevisionsList);
      setRepositories(checkedRepositories as Repository['name'][]);
    } else {
      setRevisions(displayedSelectedRevisions);
      setRepositories(selectedRevRepo as Repository['name'][]);
    }
  }, [checkedRevisionsList, selectedRevisions]);

  return (
    <Box className={styles.box} data-testid={`selected-revs-${view}`}>
      <List>
        {revisions.map((item, index) => (
          <SelectedRevisionItem
            key={item.id}
            index={index}
            item={item}
            mode={mode}
            repository={repositories[index]}
            searchType={searchType}
            isWarning={isWarning}
            view={view}
            editBtnVisible={editBtnVisible}
          />
        ))}
      </List>
    </Box>
  );
}

export default SelectedRevisions;
