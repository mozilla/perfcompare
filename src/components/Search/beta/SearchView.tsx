import PerfCompareHeader from '../../Shared/beta/PerfCompareHeader';

function SearchViewBeta() {
  return (
    <section className='perfcompare-body'>
      <PerfCompareHeader />
    </section>
  );
}

export default SearchViewBeta;
