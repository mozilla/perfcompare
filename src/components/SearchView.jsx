import * as React from 'react';

import Box from '@mui/material/Box';
import CssBaseline from '@mui/material/CssBaseline';
import Container from '@mui/material/Container';
import Typography from '@mui/material/Typography';

import SearchInput from './SearchInput';

export default function SearchView() {
  return (
    <React.Fragment>
      <CssBaseline />
      <Container maxWidth="lg">
        <Box>
          <Typography variant="h1" component="div" align="center" gutterBottom>
            PerfCompare
          </Typography>
        </Box>
        <SearchInput />
      </Container>
    </React.Fragment>
  );
}
