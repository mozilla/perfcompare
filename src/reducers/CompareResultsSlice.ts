import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import {
  fetchCompareResults,
  fetchFakeResults,
} from '../thunks/compareResultsThunk';
import type {
  CompareResultsState,
  CompareResultsItem,
  ResultsHashmap,
} from '../types/state';

const initialState: CompareResultsState = {
  data: {} as ResultsHashmap,
  loading: false,
  error: undefined,
};

function fetchPending(state: CompareResultsState) {
  state.loading = true;
  state.error = initialState.error;
}

function fetchFulfilled(
  state: CompareResultsState,
  action: PayloadAction<CompareResultsItem[]>,
) {
  const revisionHash = action.payload[0].new_rev;
  state.data[revisionHash] = action.payload;
  state.loading = initialState.loading;
}

function fetchRejected(
  state: CompareResultsState,
  action: PayloadAction<string | undefined>,
) {
  state.error = action.payload;
  state.loading = initialState.loading;
}

const compareResults = createSlice({
  name: 'compareResults',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCompareResults.pending, fetchPending)
      .addCase(fetchFakeResults.pending, fetchPending)
      .addCase(fetchCompareResults.fulfilled, fetchFulfilled)
      .addCase(fetchFakeResults.fulfilled, fetchFulfilled)
      .addCase(fetchCompareResults.rejected, fetchRejected)
      .addCase(fetchFakeResults.rejected, fetchRejected);
  },
});

export default compareResults.reducer;
