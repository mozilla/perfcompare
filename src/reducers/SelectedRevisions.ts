import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { fetchRevisionByID } from "../thunks/searchThunk";
import { Revision, SelectedRevisionsState } from "../types/state";

const initialState: SelectedRevisionsState = {
  revisions: [],
};

const selectedRevisions = createSlice({
  name: "selectedRevisions",
  initialState,
  reducers: {
    setSelectedRevisions(state, action: PayloadAction<Revision[]>) {
      state.revisions = action.payload;
    },
    deleteRevision(state, action) {
      return {
        revisions: state.revisions.filter(
          (revision) => revision.id !== action.payload
        ),
      };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchRevisionByID.fulfilled, (state, action) => {
      state.revisions = state.revisions
        .concat(action.payload[0])
        .filter(
          (revision, index, self) =>
            self.findIndex((r) => r.id === revision.id) === index
        );
    });
  },
});

export const { setSelectedRevisions, deleteRevision } =
  selectedRevisions.actions;
export default selectedRevisions.reducer;
