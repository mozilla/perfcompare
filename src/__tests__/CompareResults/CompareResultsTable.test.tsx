import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import CompareResultsTable from '../../components/CompareResults/CompareResultsTable';
import SearchView from '../../components/Search/SearchView';
import { setCompareResults } from '../../reducers/CompareResultsSlice';
import {
  addFilter,
  setFilteredResults,
  removeFilter,
} from '../../reducers/FilterCompareResultsSlice';
import { setSelectedRevisions } from '../../reducers/SelectedRevisions';
import { ActiveFilters, FilteredResults } from '../../types/types';
import getTestData from '../utils/fixtures';
import { render, renderWithRouter, store } from '../utils/setupTests';
import { screen, waitFor } from '../utils/test-utils';

const { testCompareData, paginationTestCompareData } = getTestData();

describe('Compare Results Table', () => {

  const themeMode = 'light';

  it('Should match snapshot', () => {
    render(<CompareResultsTable theme={themeMode} />);

    expect(document.body).toMatchSnapshot();
  });

  it('Should not display platform options list by default', () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);

    const platformOptionsList = screen.queryByTestId('platform-options');

    expect(platformOptionsList).toBeNull();
  });

  it('Should open platform options', async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);

    const user = userEvent.setup({ delay: null });

    await user.click(screen.getByTestId('platform-options-button'));

    const platformOptionsList = screen.getByTestId('platform-options');

    expect(platformOptionsList).toBeInTheDocument();
  });

  it('Should filter data by platform', async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);
    const rows = screen.getAllByTestId('table-row');

    expect(rows).not.toBeNull();

    act(() => {
      store.dispatch(
        addFilter({ name: 'platform', value: 'macosx1015-64-shippable-qr' }),
      );
    });

    const activeFilters: ActiveFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };
    const filteredResults: FilteredResults = {
      data: [testCompareData[0]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const filteredRows = screen.getAllByTestId('table-row');

    filteredRows.forEach((row) =>
      expect(row.firstChild).toHaveAttribute(
        'aria-label',
        'macosx1015-64-shippable-qr',
      ),
    );
  });

  it("Should filter data by confidence 'not available'", async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);
    const rows = screen.getAllByTestId('table-row');

    expect(rows).not.toBeNull();

    act(() => {
      store.dispatch(addFilter({ name: 'confidence', value: 'not available' }));
    });

    const activeFilters: ActiveFilters = {
      platform: [],
      test: [],
      confidence: ['not available'],
    };
    const filteredResults: FilteredResults = {
      data: [testCompareData[3]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const filteredRows = screen.getAllByTestId('table-row');

    filteredRows.forEach((row) =>
      expect(row.childNodes[7].firstChild).toHaveAttribute(
        'aria-label',
        'Confidence not available',
      ),
    );
  });

  it("Should check confidence 'not available'", async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);
    const user = userEvent.setup({ delay: null });

    await user.click(screen.getByTestId('confidence-options-button'));

    const option = await waitFor(() =>
      screen.getByTestId('not available-checkbox'),
    );

    expect(option).not.toBeNull();

    await user.click(option);

    expect(
      screen
        .getByTestId('not available-checkbox')
        .classList.contains('Mui-checked'),
    ).toBe(true);
  });

  it("Should filter data on 'apply' button", async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);
    const user = userEvent.setup({ delay: null });

    await user.click(screen.getByTestId('confidence-options-button'));
    await user.click(screen.getByTestId('not available-checkbox'));
    await user.click(screen.getAllByTestId('apply-filter')[0]);

    const filteredRows = screen.getAllByTestId('table-row');

    filteredRows.forEach((row) =>
      expect(row.childNodes[7].firstChild).toHaveAttribute(
        'aria-label',
        'Confidence not available',
      ),
    );
  });

  it("Should uncheck confidence 'not available'", async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);
    const user = userEvent.setup({ delay: null });

    await user.click(screen.getByTestId('confidence-options-button'));
    await user.click(screen.getByTestId('not available-checkbox'));
    await user.click(screen.getByTestId('not available-checkbox'));

    expect(
      screen
        .getByTestId('not available-checkbox')
        .classList.contains('Mui-checked'),
    ).toBe(false);
  });

  it('Should remove platform macosx1015-64-shippable-qr from filter', async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);

    act(() => {
      store.dispatch(
        addFilter({ name: 'platform', value: 'macosx1015-64-shippable-qr' }),
      );
    });

    expect(
      store.getState().filterCompareResults.activeFilters.platform,
    ).toStrictEqual(['macosx1015-64-shippable-qr']);

    const activeFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };

    let filteredResults: FilteredResults = {
      data: [testCompareData[0]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      store.dispatch(
        removeFilter({ name: 'platform', value: 'macosx1015-64-shippable-qr' }),
      );
    });

    filteredResults = {
      data: [],
      activeFilters,
      isFiltered: false,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    expect(
      store.getState().filterCompareResults.activeFilters.platform,
    ).toStrictEqual([]);
  });

  it('Should display success message for filtering data.', () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);

    const activeFilters: ActiveFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };
    const filteredResults: FilteredResults = {
      data: [testCompareData[0]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const message = screen.queryByText('Filters have been applied.');

    expect(message).toBeInTheDocument();
  });

  it('Should display no data message for filtering data.', () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);

    const activeFilters: ActiveFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: ['med'],
    };
    const filteredResults: FilteredResults = {
      data: [],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const message = screen.queryByText('No data for this combination.');

    expect(message).toBeInTheDocument();
  });

  it('Should display changed options message for filtering data.', () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);

    act(() => {
      store.dispatch(
        addFilter({ name: 'platform', value: 'macosx1015-64-shippable-qr' }),
      );
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const message = screen.queryByText(
      'Filter options have changed. Apply the new changes.',
    );

    expect(message).toBeInTheDocument();
  });

  it('Should display active options chip.', () => {
    // set compare data
    store.dispatch(setCompareResults(testCompareData));

    const { rerender } = render(<CompareResultsTable theme={themeMode} />);

    const activeFilters: ActiveFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };
    const filteredResults: FilteredResults = {
      data: [testCompareData[0]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    act(() => {
      rerender(<CompareResultsTable theme={themeMode} />);
    });

    const chip = screen.queryByLabelText('macosx1015-64-shippable-qr');

    expect(chip).toBeInTheDocument();
  });

  it("Should reset filter when 'clear' button is clicked", async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);
    const user = userEvent.setup({ delay: null });

    const rows = await waitFor(() => screen.getAllByTestId('table-row'));

    expect(rows.length).toBe(4);

    await user.click(screen.getByTestId('platform-options-button'));
    await user.click(screen.getByTestId('macosx1015-64-shippable-qr-checkbox'));
    await user.click(screen.getAllByTestId('apply-filter')[0]);

    let filteredRows = await waitFor(() => screen.getAllByTestId('table-row'));

    expect(filteredRows.length).toBe(1);

    await user.click(screen.getByTestId('clear-filter'));

    filteredRows = await waitFor(() => screen.getAllByTestId('table-row'));

    expect(filteredRows.length).toEqual(rows.length);
    expect(
      store.getState().filterCompareResults.activeFilters.platform,
    ).toStrictEqual([]);
  });

  it('Should reset filter when new data is compared', async () => {
    // set results data
    store.dispatch(setCompareResults(testCompareData));

    const activeFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };

    const filteredResults: FilteredResults = {
      data: [testCompareData[0]],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });
    const { testData } = getTestData();

    act(() => {
      store.dispatch(setSelectedRevisions(testData.slice(0, 2)));
    });

    const { history } = renderWithRouter(<SearchView />);
    expect(history.location.pathname).toEqual('/');

    const user = userEvent.setup({ delay: null });

    const compareButton = document.querySelector('.compare-button');
    await user.click(compareButton as HTMLElement);

    expect(history.location.pathname).toEqual('/compare-results');
    expect(
      store.getState().filterCompareResults.activeFilters.platform,
    ).toStrictEqual([]);
    expect(store.getState().filterCompareResults.isFiltered).toStrictEqual(
      false,
    );
  });

  it('Should close filter popper', async () => {
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    // set compare data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);

    const filterButton = screen.getByTestId('platform-options-button');
    await user.click(filterButton);
    let popper = await waitFor(() => screen.getByRole('tooltip'));
    expect(popper).toBeVisible();

    const windowsCheckbox = await waitFor(() =>
      screen.getByRole('checkbox', {
        name: 'windows10-64-shippable-qr',
      }),
    );
    expect(windowsCheckbox).toBeInTheDocument();

    await user.click(windowsCheckbox);

    const applyButtons = await waitFor(() =>
      screen.getAllByRole('button', { name: 'Apply' }),
    );

    await user.click(applyButtons[1]);

    popper = await waitFor(() => screen.getByRole('tooltip'));

    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(popper).not.toBeInTheDocument();
  });

  it('Should close filter popper by clicking outside of it', async () => {
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });

    // set compare data
    store.dispatch(setCompareResults(testCompareData));

    render(<CompareResultsTable theme={themeMode} />);

    const filterButtonPlatform = await waitFor(() =>
      screen.getByTestId('platform-options-button'),
    );
    await user.click(filterButtonPlatform);

    const filterPlatfomList = screen.getByTestId('platform-options');
    let popper = await waitFor(() => screen.getByRole('tooltip'));
    expect(popper).toContainElement(filterPlatfomList);

    const filterButtonConfidence = await waitFor(() =>
      screen.getByTestId('confidence-options-button'),
    );
    expect(filterButtonConfidence).toBeInTheDocument();

    await user.click(filterButtonConfidence);

    const filterConfidenceList = await waitFor(() =>
      screen.getByTestId('confidence-options'),
    );

    popper = await waitFor(() => screen.getByRole('tooltip'));

    expect(popper).toContainElement(filterConfidenceList);
  });

  it('should reset to first table page when filtering is changed', async () => {
    // set results data
    store.dispatch(setCompareResults(paginationTestCompareData));

    const activeFilters: ActiveFilters = {
      platform: ['macosx1015-64-shippable-qr'],
      test: [],
      confidence: [],
    };

    const filteredResults: FilteredResults = {
      data: [...paginationTestCompareData.slice(1, 33)],
      activeFilters,
      isFiltered: true,
    };

    act(() => {
      store.dispatch(setFilteredResults(filteredResults));
    });

    render(<CompareResultsTable theme={themeMode} />);

    let firstPageButton = await waitFor(() =>
      screen.getByLabelText('first page'),
    );

    expect(firstPageButton).toHaveAttribute('disabled');

    const user = userEvent.setup({ delay: null });

    // goes to second page
    await user.click(screen.getByLabelText('next page'));

    firstPageButton = await waitFor(() => screen.getByLabelText('first page'));

    expect(firstPageButton).not.toHaveAttribute('disabled');

    act(() => {
      store.dispatch(
        addFilter({ name: 'test', value: 'content-process-startup' }),
      );
    });

    firstPageButton = await waitFor(() => screen.getByLabelText('first page'));

    expect(firstPageButton).toHaveAttribute('disabled');
  });
});
