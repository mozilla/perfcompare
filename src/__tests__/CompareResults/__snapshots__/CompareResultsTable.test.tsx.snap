// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compare Results Table Should match snapshot 1`] = `
<body>
  <div />
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-pfrgnn-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="a dense table"
        class="MuiTable-root css-1nmx7wt-MuiTable-root"
      />
    </div>
  </div>
</body>
`;
